src/crypto_provider.rs

```diff
@@ -269,6 +309,24 @@ pub extern "C" fn rustls_supported_ciphersuites_free(
     }
 }
 
+box_castable! {
+    /// A signing key that can be used to construct a certified key.
+    // Note(XXX): we box cast an arc over the dyn trait per the pattern described
+    //   in our docs[0] for dynamically sized types.
+    //   [0]: <https://github.com/rustls/rustls-ffi/blob/main/CONTRIBUTING.md#dynamically-sized-types>
+    pub struct rustls_signing_key(Arc<dyn SigningKey>);
+}
+
+impl rustls_signing_key {
+    /// Frees the `rustls_signing_key`. This is safe to call with `NULL`, or multiple times.
```

Comment:
"multiple times" not sure that is correct? a second call would be a UAF

Reply:
Oof. Good catch, thank you. Fixed up.

Reply:
I think there are a few other instances of this remaining.

Reply:
Argh, sorry about that. I should have done a self-review pass. 

I think I've caught the remaining instances.